src/app/services/xml-parsers/witnesses-parser.service.ts

```diff
@@ -0,0 +1,120 @@
+        groups: [],
+        encodingStructure: [],
+      },
+    };
+    Array.from(document.querySelectorAll<XMLElement>(this.tagNamesMap.witListTagName))
+      .map((list) => {
+        !isNestedInElem(list, list.tagName) ? this.parseWitnesses(list, parsedList) : this.parseWitnessesGroups(list, parsedList);
+      });
+
+    return parsedList;
+  }
+
+  private parseWitnesses(list: XMLElement, parsedList: WitnessesList) {
+    list.querySelectorAll<XMLElement>(this.tagNamesMap.witTagName)
+      .forEach((wit) => {
+        const parsedWit = this.parseWitness(wit);
+        parsedList.witnesses[parsedWit.id] = parsedWit;
+        parsedList.indexes.witnesses.push(wit.getAttribute('xml:id'));
+        if (parsedWit.group === undefined) {
+          parsedList.indexes.encodingStructure.push(wit.getAttribute('xml:id'));
+        }
+        if (parsedWit.group && !parsedList.indexes.encodingStructure.includes(parsedWit.group)) {
+          parsedList.indexes.encodingStructure.push(parsedWit.group);
+        }
+      });
+  }
+
+  private parseWitness(wit: XMLElement): Witness {
+    return {
+      id: wit.getAttribute('xml:id') || xpath(wit),
+      attributes: this.genericParserService.parseAttributes(wit),
+      content: this.parseWitnessContent(wit),
+      group: this.parseParentGroupId(wit),
+    };
+  }
+
+  private parseWitnessContent(wit: XMLElement): Description {
+    return Array.from(wit.childNodes)
+      .map((child: XMLElement) => this.genericParserService.parse(child));
+  }
+
+  private parseWitnessesGroups(list: XMLElement, parsedList: WitnessesList) {
+    const parsedGroup = this.parseWitnessGroup(list);
+    parsedList.originalGroup[parsedGroup.id] = parsedGroup;
+    parsedList.indexes.groups.push(parsedGroup.id);
+  }
+
+  private parseWitnessGroup(list: XMLElement): WitnessGroup {
+    return {
+      id: list.getAttribute('xml:id') || xpath(list),
+      name: this.parseGroupName(list) || replaceNotWordChar(list.getAttribute('xml:id')) || xpath(list),
+      attributes: this.genericParserService.parseAttributes(list),
+      content: this.parseGroupContent(list),
+      group: this.parseParentGroupId(list) || undefined,
+    };
+  }
+
+  private parseGroupName(list: XMLElement) {
+    const groupTagName = 'head';
+    const groupEl = Array.from(list.children)
+      .find((child) => child.nodeName === groupTagName);
+
+    return groupEl ? groupEl.textContent : undefined;
+  }
+
+  private parseGroupContent(list: XMLElement): Description {
+    const content = [];
+    Array.from(list.children)
+      .filter(() => Object.keys(this.tagNamesMap))
+      .map((child) => content.push(child.getAttribute('xml:id')));
+
+    return content;
+  }
+
+  private parseParentGroupId(element: XMLElement) {
+    const parentEl = element.parentElement;
+
+    if (parentEl.tagName !== this.tagNamesMap.witListTagName) {
+      this.parseParentGroupId(parentEl);
+    }
+
+    if (isNestedInElem(parentEl, parentEl.tagName)) {
```

Comment:
You just found the parent walking up the parent-child structure, shouldn't it be always true?

Reply:
I need this condition because in an encoding like below, the first `<listWit>` element, which includes all the others, should not be considered a group.
```
<listWit>
  <listWit xml:id="rec1_wits"></listWit>
  <listWit xml:id="rec2_wits"></listWit>
  <listWit xml:id="rec3_wits"></listWit>
<listWit>
```